core/src/main/java/fr/sncf/osrd/envelope/EnvelopePart.java

```diff
@@ -1,23 +1,77 @@
 package fr.sncf.osrd.envelope;
 
 
+import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
 import java.util.Arrays;
 import java.util.Objects;
 
 public final class EnvelopePart {
-    public final EnvelopeType type;
+    /** Tracks what created this envelope part */
+    public final EnvelopeSource source;
+    /** The shape of this envelope part */
+    public final EnvelopeAttitude attitude;
+
+    /** Whether the train can actually follow this envelope part
+     * TODO(Giuliana): find a better name
+     */
+    public final boolean physicallyAccurate;
+
+    /** A list of N spacial offsets */
     public final double[] positions;
+    /** A list of N speeds, one per position */
     public final double[] speeds;
-    public final boolean isCoasting;
 
-    /** Create an EnvelopePart */
-    public EnvelopePart(EnvelopeType type, double[] positions, double[] speeds, boolean isCoasting) {
+    /** A list of N - 1 time deltas between positions */
+    private double[] cachedTimes;
+
+    /** Creates an EnvelopePart */
+    @SuppressFBWarnings({"EI_EXPOSE_REP2"})
+    public EnvelopePart(
+            EnvelopeSource source,
+            EnvelopeAttitude attitude,
+            boolean physicallyAccurate,
+            double[] positions,
+            double[] speeds,
+            double[] cachedTimes
+    ) {
         assert positions.length == speeds.length;
         assert positions.length >= 2;
-        this.type = type;
+        assert cachedTimes == null || cachedTimes.length == positions.length - 1;
+        this.source = source;
+        this.attitude = attitude;
+        this.physicallyAccurate = physicallyAccurate;
+
         this.positions = positions;
         this.speeds = speeds;
-        this.isCoasting = isCoasting;
+        this.cachedTimes = cachedTimes;
+    }
+
+    public EnvelopePart(
+            EnvelopeSource source,
+            EnvelopeAttitude attitude,
+            boolean physicallyAccurate,
+            double[] positions,
+            double[] speeds
+    ) {
+        this(source, attitude, physicallyAccurate, positions, speeds, null);
+    }
+
+    private static double[] computeTimes(double[] positions, double[] speeds) {
```

Comment:
computeTime does a quite bad approximation. Moreover delta time is constant in our case. Do we really need this method?

Reply:
acc = ( v2^2 - v1^2 ) / ( 2*( x2 - x1 ))
deltat = deltav / acc